### PR TITLE
Add quotaCenter, rateLimiter and rateCollector

### PR DESCRIPTION
**Pull request by bigsheeper**
Signed-off-by: bigsheeper <yihao.dai@zilliz.com>

issues: #18635

### CODE REVIEW

File: configs/milvus.yaml [hunk 1]
```diff
@@ -338,3 +338,66 @@ common:
...
+    queryRate: 0.01 # MB/s
+    maxSearchSize: 0.1 # the maximum size(MB) of a single search request that allowed.
+    maxQuerySize: 0.01 # the maximum size(MB) of a single query request that allowed.
+
+  limits:
+    maxCollectionNum: 32768 # max number of collections in a Milvus instance.
+    # TODO: add more limits
+
+  # forceDenyWriting decides whether dml requests are allowed.
+  forceDenyWriting:
```
Comment (czs007) [inline configs/milvus.yaml:372]: better to rename as "limitWritting" and rename enable to "forceDenny"

File: internal/metrics/quota_metrics.go [hunk 2]
```diff
@@ -0,0 +1,247 @@
...
+
+	"github.com/prometheus/client_golang/prometheus"
+	"go.uber.org/zap"
+
+	"github.com/milvus-io/milvus/internal/log"
+	"github.com/milvus-io/milvus/internal/proto/internalpb"
+	"github.com/milvus-io/milvus/internal/util/typeutil"
+)
+
+// limiter's rates
```
Comment (czs007) [inline internal/metrics/quota_metrics.go:30]: Please spread these out into the metrics module of each role.

File: internal/metrics/quota_metrics.go [hunk 3]
```diff
@@ -0,0 +1,247 @@
...
+	//		Namespace: milvusNamespace,
+	//		Subsystem: typeutil.ProxyRole,
+	//		Name:      "query_rate",
+	//		Help:      "",
+	//	}, []string{nodeIDLabelName})
+)
+
+// nodes rates
+var (
+	DataNodeConsumeRate = prometheus.NewCounterVec(
```
Comment (czs007) [inline internal/metrics/quota_metrics.go:134]: bettter to rename as DataNodeConsumeCounter

File: internal/distributed/proxy/client/client.go [hunk 4]
```diff
@@ -214,3 +214,31 @@ func (c *Client) RefreshPolicyInfoCache(ctx context.Context, req *proxypb.Refres
 	}
 	return ret.(*commonpb.Status), err
 }
+
+// GetMetrics gets the metrics of the proxy.
+func (c *Client) GetMetrics(ctx context.Context, req *milvuspb.GetMetricsRequest) (*milvuspb.GetMetricsResponse, error) {
```
Comment (czs007) [inline internal/distributed/proxy/client/client.go:219]: better to rename as GetProxyMetrics   to  differentiate from GetMetrics

File: internal/distributed/proxy/service.go [hunk 5]
```diff
@@ -162,6 +162,14 @@ func (s *Server) startExternalGrpc(grpcPort int, errChan chan error) {
 	}
 	log.Debug("Proxy server already listen on tcp", zap.Int("port", grpcPort))
 
+	multiLimiter := proxy.NewMultiRateLimiter()
```
Comment (czs007) [inline internal/distributed/proxy/service.go:165]: Create RateLimiter in internal/proxy. 
and Add GetRateLimiter to  proxy interface

File: internal/proto/internal.proto [hunk 6]
```diff
@@ -322,3 +322,19 @@ message ShowConfigurationsResponse {
   common.Status status = 1;
   repeated common.KeyValuePair configuations = 2; 
 }
+
+enum RateType {
```
Comment (czs007) [inline internal/proto/internal.proto:326]: add DMLBulkLoad
add DDLFlush DDLCompaction ....

File: internal/util/metricsinfo/topology.go [hunk 7]
```diff
@@ -23,7 +24,12 @@ import (
 
 // ConstructComponentName returns a name according to the role name and its' id
 func ConstructComponentName(role string, id typeutil.UniqueID) string {
-	return role + strconv.Itoa(int(id))
+	return role + "-" + strconv.Itoa(int(id))
+}
+
+// GetRoleNameByComponentName returns role name by component name.
+func GetRoleNameByComponentName(componentName string) string {
```
Comment (czs007) [inline internal/util/metricsinfo/topology.go:31]: delete this and add role member in response.

File: internal/util/metricsinfo/quota_metric.go [hunk 8]
```diff
@@ -0,0 +1,52 @@
...
+	UsedMem  uint64
+	TotalMem uint64
+}
+
+type QueryNodeQuotaMetrics struct {
+	NodeID          typeutil.UniqueID
+	Mm              MemMetric
+	MinTSafe        typeutil.Timestamp
+	SearchNQInQueue int64
+	QueriesInQueue  int64
```
Comment (czs007) [inline internal/util/metricsinfo/quota_metric.go:33]: add more metrics

File: internal/rootcoord/quota_center.go [hunk 9]
```diff
@@ -0,0 +1,408 @@
...
+}
+
+// calculateRates calculates target rates by different strategies.
+func (q *QuotaCenter) calculateRates() error {
+	q.resetCurrentRates()
+
+	if Params.QuotaConfig.ForceDenyWriting {
+		q.forceDenyWriting()
+	} else {
+		percentage, err := q.tSafeDelayed()
```
Comment (czs007) [inline internal/rootcoord/quota_center.go:247]: check datanode timetick

File: internal/querynode/rate_collector.go [hunk 10]
```diff
@@ -0,0 +1,80 @@
...
+	tSafes          map[Channel]Timestamp
+}
+
+func newRateCollector() *rateCollector {
+	return &rateCollector{
+		tSafes: make(map[Channel]Timestamp),
+	}
+}
+
+func (r *rateCollector) addNQ(nq int64) {
```
Comment (czs007) [inline internal/querynode/rate_collector.go:40]: change this

File: internal/querynode/impl.go [hunk 11]
```diff
@@ -923,6 +935,11 @@ func (node *QueryNode) Query(ctx context.Context, req *querypb.QueryRequest) (*i
 		zap.Uint64("guaranteeTimestamp", req.Req.GetGuaranteeTimestamp()),
 		zap.Uint64("timeTravel", req.GetReq().GetTravelTimestamp()))
 
+	if !req.FromShardLeader {
```
Comment (czs007) [inline internal/querynode/impl.go:938]: change this

**Comment by bigsheeper** [comment]
rerun ut

File: internal/datanode/metrics_info.go [hunk 12]
```diff
@@ -23,9 +23,40 @@ import (
 	"github.com/milvus-io/milvus/internal/proto/internalpb"
 	"github.com/milvus-io/milvus/internal/proto/milvuspb"
 	"github.com/milvus-io/milvus/internal/util/metricsinfo"
+	"github.com/milvus-io/milvus/internal/util/ratecollector"
 	"github.com/milvus-io/milvus/internal/util/typeutil"
 )
 
+// getQuotaMetrics returns DataNodeQuotaMetrics.
+func getQuotaMetrics() (*metricsinfo.DataNodeQuotaMetrics, error) {
+	var err error
```
Comment (xiaofan-luan) [inline internal/datanode/metrics_info.go:32]: add unflushed data size.

File: internal/proxy/multi_rate_limiter.go [hunk 13]
```diff
@@ -0,0 +1,141 @@
...
+	log.Debug("RateLimiter setRates", zap.Any("rates", rates))
+}
+
+// registerLimiters register limiter for all rate types.
+func (rl *rateLimiter) registerLimiters() {
+	for rt := range internalpb.RateType_name {
+		var r float64
+		var b int
+		switch internalpb.RateType(rt) {
+		case internalpb.RateType_DDLCollection:
```
Comment (xiaofan-luan) [inline internal/proxy/multi_rate_limiter.go:104]: all the DDL related limiter should be on rootcoord?

File: internal/util/paramtable/quota_param.go [hunk 14]
```diff
@@ -0,0 +1,204 @@
...
+	p.DMLInsertRate = megaBytesRate2Bytes(rate)
+}
+
+func (p *quotaConfig) initDMLDeleteRate() {
+	rate := p.Base.ParseFloatWithDefault("quotaAndLimits.dml.deleteRate", 1)
+	p.DMLDeleteRate = megaBytesRate2Bytes(rate)
+}
+
+func (p *quotaConfig) initDMLBulkLoadRate() {
+	p.DMLBulkLoadRate = p.Base.ParseFloatWithDefault("quotaAndLimits.dml.deleteRate", 1)
```
Comment (xiaofan-luan) [inline internal/util/paramtable/quota_param.go:140]: wrong token

Comment (xiaofan-luan) [reply internal/util/paramtable/quota_param.go:140]: should be limited by throughput

File: internal/util/paramtable/quota_param.go [hunk 15]
```diff
@@ -0,0 +1,204 @@
...
+}
+
+func (p *quotaConfig) initDQLSearchRate() {
+	rate := p.Base.ParseFloatWithDefault("quotaAndLimits.dql.searchRate", 0.1)
```
Comment (xiaofan-luan) [inline internal/util/paramtable/quota_param.go:144]: support to limit by search/query result size.
not search size

File: internal/util/metricsinfo/quota_metric.go [hunk 16]
```diff
@@ -0,0 +1,74 @@
...
+
+import (
+	"github.com/milvus-io/milvus/internal/util/typeutil"
+)
+
+// RateMetricLabel defines the metric label collected from nodes.
+type RateMetricLabel = string
+
+const (
+	NQPerSecond             RateMetricLabel = "NQPerSecond"
```
Comment (xiaofan-luan) [inline internal/util/metricsinfo/quota_metric.go:27]: 1. search query queue latency
2. search/query result throughput

**Comment by xiaofan-luan** [comment]
by default:
We enable 
1. Memory Protection 
2. TT Protection 

User can enable:
1. Insert/Delete/BulkLoad Throughput limitation
2. Search/Query Result NQ limitation
3. Search/Query Queue Time Limitation

File: internal/util/paramtable/quota_param.go [hunk 17]
```diff
@@ -0,0 +1,260 @@
...
+	p.DDLCompactionRate = p.Base.ParseFloatWithDefault("quotaAndLimits.ddl.compactionRate", -1)
+}
+
+func megaBytesRate2Bytes(f float64) float64 {
+	return f * 1024 * 1024
+}
+
+func (p *quotaConfig) checkMinMax(min, max float64) {
+	if min > 0 && max > 0 && min > max {
+		panic(fmt.Errorf("init QuotaConfig failed, max rate must be greater than or equal to min rate"))
```
Comment (czs007) [inline internal/util/paramtable/quota_param.go:142]: min =3  
max = -1 
will panic

File: internal/util/ratelimitutil/limiter.go [hunk 18]
```diff
@@ -0,0 +1,142 @@
...
+
+package ratelimitutil
+
+import (
+	"math"
+	"sync"
+	"time"
+)
+
+// Partial implementations refer to https://github.com/golang/time
```
Comment (czs007) [inline internal/util/ratelimitutil/limiter.go:25]: https://github.com/golang/time/blob/master/rate/rate.go

File: internal/proxy/rate_limit_interceptor.go [hunk 19]
```diff
@@ -0,0 +1,85 @@
...
+)
+
+// RateLimitInterceptor returns a new unary server interceptors that performs request rate limiting.
+func RateLimitInterceptor(limiter types.Limiter) grpc.UnaryServerInterceptor {
+	return func(ctx context.Context, req interface{}, info *grpc.UnaryServerInfo, handler grpc.UnaryHandler) (interface{}, error) {
+		rt, n, err := getRequestInfo(req)
+		if err == nil {
+			limit, rate := limiter.Limit(rt, n)
+			if rate == 0 {
+				return nil, status.Errorf(codes.ResourceExhausted, commonpb.ErrorCode_ForceDeny.String())
```
Comment (czs007) [inline internal/proxy/rate_limit_interceptor.go:42]: better to  return error code

File: internal/querynode/task_read.go [hunk 20]
```diff
@@ -64,6 +65,9 @@ type baseReadTask struct {
 	step               TaskStep
 	queueDur           time.Duration
 	reduceDur          time.Duration
+	waitTsDur          time.Duration
+	waitTSTrOnce       sync.Once
```
Comment (czs007) [inline internal/querynode/task_read.go:69]: remove once

File: internal/util/ratelimitutil/rate_collector.go [hunk 21]
```diff
@@ -0,0 +1,172 @@
...
+
+// shift slides the window per granularity.
+func (r *RateCollector) shift() {
+	ticker := time.NewTicker(r.granularity)
+	defer ticker.Stop()
+	for {
+		select {
+		case <-r.stopChan:
+			return
+		case <-ticker.C:
```
Comment (czs007) [inline internal/util/ratelimitutil/rate_collector.go:161]: better to remove this

**Comment by bigsheeper** [comment]
rerun ut

**Pull request merged**